File: app/src/test/java/com/google/android/stardroid/math/GeometryTest.java
// Copyright 2008 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.android.stardroid.math;

import junit.framework.TestCase;

public class GeometryTest extends TestCase {
  private final double TOL = 0.00001;

  public void assertVectorsEqual(Vector3 v, Vector3 w) {
    assertEquals(v.x, w.x, TOL);
    assertEquals(v.y, w.y, TOL);
    assertEquals(v.z, w.z, TOL);
  }

  private void assertMatrixSame(Matrix3x3 m1, Matrix3x3 m2, double tol) {
    assertEquals(m1.xx, m2.xx, tol);
    assertEquals(m1.xy, m2.xy, tol);
    assertEquals(m1.xz, m2.xz, tol);
    assertEquals(m1.yx, m2.yx, tol);
    assertEquals(m1.yy, m2.yy, tol);
    assertEquals(m1.yz, m2.yz, tol);
    assertEquals(m1.zx, m2.zx, tol);
    assertEquals(m1.zy, m2.zy, tol);
    assertEquals(m1.zz, m2.zz, tol);
  }

  private float[][] allTestValues = { {0, 0, 1, 0, 0},
                                      {90, 0, 0, 1, 0},
                                      {0, 90, 0, 0, 1},
                                      {180, 0, -1, 0, 0},
                                      {0, -90, 0, 0, -1},
                                      {270, 0, 0, -1, 0} };

  public void testSphericalToCartesians() {
    for (float[] testValues : allTestValues) {
      float ra = testValues[0];
      float dec = testValues[1];
      float x = testValues[2];
      float y = testValues[3];
      float z = testValues[4];
      Vector3 result = Geometry.getXYZ(new RaDec(ra, dec));
      assertEquals(x, result.x, TOL);
      assertEquals(y, result.y, TOL);
      assertEquals(z, result.z, TOL);
    }
  }

  public void testCartesiansToSphericals() {
    for (float[] testValues : allTestValues) {
      float ra = testValues[0];
      float dec = testValues[1];
      float x = testValues[2];
      float y = testValues[3];
      float z = testValues[4];
      RaDec result =
          RaDec.fromGeocentricCoords(new Vector3(x, y, z));
      assertEquals(ra, result.getRa(), TOL);
      assertEquals(dec, result.getDec(), TOL);
    }
  }

  public void testVectorProduct() {
    // Check that z is x X y
    Vector3 x = new Vector3(1, 0, 0);
    Vector3 y = new Vector3(0, 1, 0);
    Vector3 z = Geometry.vectorProduct(x, y);
    assertVectorsEqual(z, new Vector3(0, 0, 1));

    // Check that a X b is perpendicular to a and b
    Vector3 a = new Vector3(1, -2, 3);
    Vector3 b = new Vector3(2, 0, -4);
    Vector3 c = Geometry.vectorProduct(a, b);
    double aDotc = Geometry.scalarProduct(a, c);
    double bDotc = Geometry.scalarProduct(b, c);
    assertEquals(0, aDotc, TOL);
    assertEquals(0, bDotc, TOL);

    // Check that |a X b| is correct
    Vector3 v= new Vector3(1, 2, 0);
    Vector3 ww = Geometry.vectorProduct(x, v);
    float wwDotww = Geometry.scalarProduct(ww, ww);
    assertEquals(Math.pow(1f * Math.sqrt(5) * Math.sin(Math.atan(2)),2),
        wwDotww, TOL);
  }

  public void testMatrixInversion() {
    Matrix3x3 m = new Matrix3x3(1, 2, 0, 0, 1, 5, 0, 0, 1);
    System.out.println(GeometryTest.formatMatrix(m));
    Matrix3x3 inv = m.getInverse();
    System.out.println(GeometryTest.formatMatrix(inv));
    Matrix3x3 product = Geometry.matrixMultiply(m, inv);
    System.out.println(GeometryTest.formatMatrix(product));
  }

  public void testCalculateRotationMatrix() {
    Matrix3x3 noRotation = Geometry.calculateRotationMatrix(0, new Vector3(1, 2, 3));
    Matrix3x3 identity = new Matrix3x3(1, 0, 0, 0, 1, 0, 0, 0, 1);
    assertMatrixSame(identity, noRotation, TOL);

    Matrix3x3 rotAboutZ = Geometry.calculateRotationMatrix(90, new Vector3(0, 0, 1));
    assertMatrixSame(new Matrix3x3(0, 1, 0, -1, 0, 0, 0, 0, 1), rotAboutZ, TOL);

    Vector3 axis = new Vector3(2, -4, 1);
    axis.normalize();
    Matrix3x3 rotA = Geometry.calculateRotationMatrix(30, axis);
    Matrix3x3 rotB = Geometry.calculateRotationMatrix(-30, axis);

    Matrix3x3 shouldBeIdentity = Geometry.matrixMultiply(rotA, rotB);
    assertMatrixSame(identity, shouldBeIdentity, TOL);

    Vector3 axisPerpendicular = new Vector3(4, 2, 0);
    Vector3 rotatedAxisPerpendicular = Geometry.matrixVectorMultiply(rotA, axisPerpendicular);

    // Should still be perpendicular
    assertEquals(0, Geometry.scalarProduct(axis, rotatedAxisPerpendicular), TOL);
    // And the angle between them should be 30 degrees
    axisPerpendicular.normalize();
    rotatedAxisPerpendicular.normalize();
    assertEquals(Math.cos(30.0 * Geometry.DEGREES_TO_RADIANS),
                 Geometry.scalarProduct(axisPerpendicular, rotatedAxisPerpendicular),
                 TOL);
  }

  public void testMatrixMultiply() {
    Matrix3x3 m1 = new Matrix3x3(1, 2, 4, -1, -3, 5, 3, 2, 6);
    Matrix3x3 m2 = new Matrix3x3(3, -1, 4, 0, 2, 1, 2, -1, 2);
    Vector3 v1 = new Vector3(0, -1, 2);
    Vector3 v2 = new Vector3(2, -2, 3);

    Matrix3x3Test.assertMatricesEqual(
        new Matrix3x3(11, -1, 14, 7, -10, 3, 21, -5, 26),
        Geometry.matrixMultiply(m1, m2), (float) TOL);

    assertVectorsEqual(new Vector3(6, 13, 10),
          Geometry.matrixVectorMultiply(m1, v1));
    assertVectorsEqual(new Vector3(10, 19, 20),
          Geometry.matrixVectorMultiply(m1, v2));
  }

  public static String formatMatrix(Matrix3x3 m) {
    return m.xx + " " + m.xy + " " + m.xz + " "
      + m.yx + " " + m.yy + " " + m.yz + " "
      + m.zx + " " + m.zy + " " + m.zz;
  }
}

File: app/src/test/java/com/google/android/stardroid/math/GeometryTest.kt
// Copyright 2008 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package com.google.android.stardroid.math

import com.google.android.stardroid.math.Geometry.calculateRotationMatrix
import com.google.android.stardroid.math.Geometry.getXYZ
import com.google.android.stardroid.math.Geometry.matrixMultiply
import com.google.android.stardroid.math.Geometry.matrixVectorMultiply
import com.google.android.stardroid.math.Geometry.scalarProduct
import com.google.android.stardroid.math.Geometry.vectorProduct
import com.google.android.stardroid.math.Matrix3x3Test.Companion.assertMatricesEqual
import com.google.android.stardroid.math.RaDec.Companion.fromGeocentricCoords
import com.google.common.truth.Truth.assertThat
import org.junit.Test

private const val TOL = 0.00001f

class GeometryTest {
    fun assertVectorsEqual(v: Vector3, w: Vector3) {
        assertThat(w.x).isWithin(TOL).of(v.x)
        assertThat(w.y).isWithin(TOL).of(v.y)
        assertThat(w.z).isWithin(TOL).of(v.z)
    }

    private fun assertMatrixSame(m1: Matrix3x3, m2: Matrix3x3, tol: Float) {
        assertThat(m2.xx).isWithin(tol).of(m1.xx)
        assertThat(m2.xy).isWithin(tol).of(m1.xy)
        assertThat(m2.xz).isWithin(tol).of(m1.xz)
        assertThat(m2.yx).isWithin(tol).of(m1.yx)
        assertThat(m2.yy).isWithin(tol).of(m1.yy)
        assertThat(m2.yz).isWithin(tol).of(m1.yz)
        assertThat(m2.zx).isWithin(tol).of(m1.zx)
        assertThat(m2.zy).isWithin(tol).of(m1.zy)
        assertThat(m2.zz).isWithin(tol).of(m1.zz)
    }

    private val allTestValues = arrayOf(
        floatArrayOf(0f, 0f, 1f, 0f, 0f),
        floatArrayOf(90f, 0f, 0f, 1f, 0f),
        floatArrayOf(0f, 90f, 0f, 0f, 1f),
        floatArrayOf(180f, 0f, -1f, 0f, 0f),
        floatArrayOf(0f, -90f, 0f, 0f, -1f),
        floatArrayOf(270f, 0f, 0f, -1f, 0f)
    )

    @Test
    fun testSphericalToCartesians() {
        for (testValues in allTestValues) {
            val ra = testValues[0]
            val dec = testValues[1]
            val x = testValues[2]
            val y = testValues[3]
            val z = testValues[4]
            val (x1, y1, z1) = getXYZ(RaDec(ra, dec))
            assertThat(x1).isWithin(TOL).of(x)
            assertThat(y1).isWithin(TOL).of(y)
            assertThat(z1).isWithin(TOL).of(z)
        }
    }

    @Test
    fun testCartesiansToSphericals() {
        for (testValues in allTestValues) {
            val ra = testValues[0]
            val dec = testValues[1]
            val x = testValues[2]
            val y = testValues[3]
            val z = testValues[4]
            val (ra1, dec1) = fromGeocentricCoords(Vector3(x, y, z))
            assertThat(ra1).isWithin(TOL).of(ra)
            assertThat(dec1).isWithin(TOL).of(dec)
        }
    }

    @Test
    fun testVectorProduct() {
        // Check that z is x X y
        val x = Vector3(1f, 0f, 0f)
        val y = Vector3(0f, 1f, 0f)
        val z = vectorProduct(x, y)
        assertVectorsEqual(z, Vector3(0f, 0f, 1f))

        // Check that a X b is perpendicular to a and b
        val a = Vector3(1f, -2f, 3f)
        val b = Vector3(2f, 0f, -4f)
        val c = vectorProduct(a, b)
        val aDotc = scalarProduct(a, c)
        val bDotc = scalarProduct(b, c)
        assertThat(aDotc).isWithin(TOL).of(0.0f)
        assertThat(bDotc).isWithin(TOL).of(0.0f)

        // Check that |a X b| is correct
        val v = Vector3(1f, 2f, 0f)
        val ww = vectorProduct(x, v)
        val wwDotww = scalarProduct(ww, ww)
        assertThat(wwDotww).isWithin(TOL)
            .of(Math.pow(1f * Math.sqrt(5.0) * Math.sin(Math.atan(2.0)), 2.0).toFloat())
    }

    @Test
    fun testMatrixInversion() {
        val m = Matrix3x3(1f, 2f, 0f, 0f, 1f, 5f, 0f, 0f, 1f)
        val inv = m.inverse
        val product = matrixMultiply(m, inv!!)
        val identity = Matrix3x3(1f, 0f, 0f, 0f, 1f, 0f, 0f, 0f, 1f)
        assertMatrixSame(product, identity, TOL)
    }

    @Test
    fun testCalculateRotationMatrix() {
        val noRotation = calculateRotationMatrix(0f, Vector3(1f, 2f, 3f))
        val identity = Matrix3x3(1f, 0f, 0f, 0f, 1f, 0f, 0f, 0f, 1f)
        assertMatrixSame(identity, noRotation, TOL)
        val rotAboutZ = calculateRotationMatrix(90f, Vector3(0f, 0f, 1f))
        assertMatrixSame(Matrix3x3(0f, 1f, 0f, -1f, 0f, 0f, 0f, 0f, 1f), rotAboutZ, TOL)
        val axis = Vector3(2f, -4f, 1f)
        axis.normalize()
        val rotA = calculateRotationMatrix(30f, axis)
        val rotB = calculateRotationMatrix(-30f, axis)
        val shouldBeIdentity = matrixMultiply(rotA, rotB)
        assertMatrixSame(identity, shouldBeIdentity, TOL)
        val axisPerpendicular = Vector3(4f, 2f, 0f)
        val rotatedAxisPerpendicular = matrixVectorMultiply(rotA, axisPerpendicular)

        // Should still be perpendicular
        assertThat(scalarProduct(axis, rotatedAxisPerpendicular)).isWithin(TOL).of(0.0f)
        // And the angle between them should be 30 degrees
        axisPerpendicular.normalize()
        rotatedAxisPerpendicular.normalize()
        assertThat(scalarProduct(axisPerpendicular, rotatedAxisPerpendicular)).isWithin(TOL)
            .of(
                Math.cos(30.0 * Geometry.DEGREES_TO_RADIANS).toFloat()
            )
    }

    @Test
    fun testMatrixMultiply() {
        val m1 = Matrix3x3(1f, 2f, 4f, -1f, -3f, 5f, 3f, 2f, 6f)
        val m2 = Matrix3x3(3f, -1f, 4f, 0f, 2f, 1f, 2f, -1f, 2f)
        val v1 = Vector3(0f, -1f, 2f)
        val v2 = Vector3(2f, -2f, 3f)
        assertMatricesEqual(
            Matrix3x3(11f, -1f, 14f, 7f, -10f, 3f, 21f, -5f, 26f),
            matrixMultiply(m1, m2), TOL.toFloat()
        )
        assertVectorsEqual(
            Vector3(6f, 13f, 10f),
            matrixVectorMultiply(m1, v1)
        )
        assertVectorsEqual(
            Vector3(10f, 19f, 20f),
            matrixVectorMultiply(m1, v2)
        )
    }
}